File: __tests__/fixtures/kernel-parity/torture.kt
/** KDoc for the file's package. */
package com.example.torture

import com.example.other.OtherClass
import com.example.util.helper
import com.example.wild.*
import com.example.alias.LongName as Short

// line comment run 1
// line comment run 2
fun topLevel(x: Int, s: String): WidgetK {
    val local = x + 1
    return WidgetK(local)
}

/** KDoc on extension fn. */
fun WidgetK.extend(n: Int): Int {
    render()
    return n
}

fun <T> List<T>.genericExt(): T = first()

fun com.example.Qualified.qext() {}

suspend fun suspender(): Unit { helper() }

private internal fun visFn() {}

fun inferred() = helper()

fun nullableRet(): WidgetK? = null

fun lambdaRet(): (Int) -> Unit = { }

expect fun platformThing(): Int

actual fun actualThing(): Int = 1

tailrec fun tailer(n: Int): Int = if (n <= 0) 0 else tailer(n - 1)

infix fun Int.pow(e: Int): Int = this

operator fun WidgetK.plus(o: WidgetK): WidgetK = this

val topVal: Int = 3
var topVar = "s"
const val TOP_CONST = 99
val topDelegated by lazy { WidgetK(1) }
val (destA, destB) = makePair()
val withGetter: Int
    get() = 42

class WidgetK(val size: Int, private var name: String = defaultName()) {
    val area: Int = size * size
    var label: String? = null
    val computed: Int
        get() = size * 2

    init {
        val initLocal = 5
        register(initLocal)
    }

    constructor(s: String) : this(s.length) {
        log(s)
    }

    fun render(): Unit {
        draw(size)
    }

    fun chainInner(): WidgetK = this

    companion object {
        val SHARED = WidgetK(0)
        const val COMPANION_CONST = 7
        fun create(): WidgetK = WidgetK(1)
    }

    companion object Named { }
}

data class DataK(val a: Int, val b: String)

abstract class AbstractK {
    abstract fun impl(): Int
}

open class OpenBase(n: Int) {
    open fun over() {}
}

class SubK(n: Int) : OpenBase(n), Drawable, Comparable<SubK> {
    override fun over() {}
    override fun compareTo(other: SubK): Int = 0
    override fun draw() {}
}

class QualifiedSuper : com.example.deep.RemoteBase() { }

class DelegatedImpl(d: Drawable) : Drawable by d

interface Drawable {
    fun draw()
    fun outline(): Int = 1
    val prop: Int get() = 2
}

sealed class SealedOp {
    object Add : SealedOp()
    data class Mul(val f: Int) : SealedOp()
}

sealed interface SealedIface

enum class Color {
    RED, GREEN, BLUE
}

enum class Http(val code: Int) {
    OK(200) {
        override fun label(): String = "ok"
    },
    ERR(500) {
        override fun label(): String = "err"
    };

    abstract fun label(): String
    fun common(): Int = code
    companion object {
        fun of(c: Int): Http = OK
    }
}

object Registry {
    val instances = mutableListOf<WidgetK>()
    var count = 0
    const val REG_CONST = 1
    fun register(w: WidgetK) { instances.add(w) }
}

annotation class MyMarker(val why: String = "")

@MyMarker
class Annotated {
    @JvmStatic
    fun jvmStatic() {}

    @Deprecated("gone", ReplaceWith("new"))
    fun old() {}

    @field:JvmField
    val fielded: Int = 1

    @get:MyMarker
    val got: Int = 2
}

typealias Handler = (Int) -> Unit
typealias WidgetList = List<WidgetK>

expect class PlatformFile {
    fun path(): String
}

actual class ActualFile {
    actual fun path(): String = "/"
}

actual typealias PlatformClock = java.time.Clock

fun caller() {
    val w = WidgetK(1)
    w.render()
    this.toString()
    super.hashCode()
    Registry.register(w)
    Registry.count
    Color.RED
    com.example.Fq.CONST_READ
    WidgetK.create().render()
    Foo.getInstance().bar()
    lowerFactory().chain()
    w.chainInner().render()
    "literal".uppercase()
    5.toString()
    listOf(1, 2).size
    w.label?.length
    w.label!!.length
    helper()
    Short.static()
    val fn: Handler = { i -> println(i) }
    fn(3)
    (fn)(4)
    run { helper() }
    listOf(1).forEach { it + 1 }
    w.let { it.render() }
    generic<Int>(1)
    register(::topLevel)
    register(OtherClass::handle)
    register(w::render)
    register(this::caller)
    obtain(String::class)
    val m = ::caller
    val bound = w::render
    val s = "interp $topVal and ${w.render()} end"
    val multi = """raw $topVal"""
    when (w.size) {
        1 -> helper()
        else -> draw(0)
    }
    if (topVal > 1) { helper() }
    for (i in 1..3) { draw(i) }
    fun localFn(): Int = 5
    localFn()
    class LocalClass {
        fun lm() {}
    }
    object LocalObj {
        fun om() {}
    }
    val anon = object : Drawable {
        override fun draw() { helper() }
    }
    anon.draw()
    label@ for (i in 1..2) { break@label }
    val backtick = `weird name`()
}

fun `weird name`(): Int = 1

fun trailing(block: (Int) -> Int): Int = block(1)

fun useTrailing() {
    trailing { it * 2 }
    trailing() { it * 3 }
}

fun defaults(a: Int = compute(), b: String = "x") {}

fun varargFn(vararg xs: Int) {}

fun destructuringBody(p: Pair<Int, Int>) {
    val (x, y) = p
    draw(x + y)
}

fun assignRefs() {
    Registry.count = 5
    Registry.count += 1
}

fun nullish(x: WidgetK?) {
    x?.render()
    val l = x ?: WidgetK(0)
}

fun stringsEdge() {
    val a = "quote \" and dollar ${'$'} done"
}

fun labeledLambda() {
    listOf(1).forEach loop@{ if (it == 0) return@loop }
}

fun whereClause(): Int where Int : Comparable<Int> = 1

File: __tests__/fixtures/kernel-parity/torture.scala
package com.example.torture

import com.example.other.OtherClass
import com.example.multi.{Alpha, Beta}
import com.example.wild._
import com.example.star.*
import com.example.alias.{LongName => Short}

/** Scaladoc for topLevel — kept? */
def topLevel(a: Int): WidgetS = new WidgetS(a)

// line one
// line two
def lineDoc(): Int = 1

def curried(a: Int)(b: String)(implicit ord: Ordering[Int]): Int = a

def genericDef[A: Numeric, B <: BoundT](x: A): B = ???

def inferredRet(a: Int) = a + 1

def unitRet(): Unit = ()

def genericLeak[T](t: T): T = t

def qualRet(w: WidgetS): com.example.other.Remote = ???

private def privTop(): Int = 1

@main def entry(): Unit = topLevel(1)

@deprecated("gone", "1.0") def old(): Int = 0

val topVal: Int = 3
var topVar = 4
lazy val topLazy = compute()
val (tupA, tupB) = (1, 2)
val Some(extracted) = Option(9)
val SHARED_TABLE: Map[String, Int] = Map.empty
val topInit = WidgetS.create()

class WidgetS(val size: Int, label: String = defaultLabel()) extends BaseW(size) with Drawable with Ordered[WidgetS] {
  val cachedName: String = "w"
  var mutable = 0
  private val secret = 3
  @volatile var annotatedField: Int = 0

  def this() = this(0, "d")

  def render(): RenderResult = {
    val local = helperR(size)
    Registry.register(this)
    new RenderResult(local)
  }

  def +(other: WidgetS): WidgetS = new WidgetS(size + other.size)

  def compare(that: WidgetS): Int = size - that.size

  @inline def fast(): Int = 1

  protected def hook(): Unit = ()
}

object WidgetS {
  val DEFAULT_SIZE = 10
  def create(): WidgetS = new WidgetS(DEFAULT_SIZE, "c")
  def readDefault(): Int = DEFAULT_SIZE + 1
}

case class DataS(x: Int, y: String = mkY())

case object SingletonS extends MarkerT

abstract class AbsS {
  def abstractM(): Int
  def concrete(): Int = abstractM() + 1
}

trait Drawable {
  def draw(): Unit
  def area(): Double = 0.0
  val traitVal: Int = 7
}

sealed trait Shape
object Circle extends Shape
class Square extends Shape

trait MarkerT

class DelegatedImpl(d: Drawable) extends Drawable {
  def draw(): Unit = d.draw()
}

enum Http {
  case Ok, NotFound
  case Custom(code: Int)
  def describe(): String = s"code"
}

enum Planet(mass: Double) {
  case Earth extends Planet(5.9)
  case Mars extends Planet(0.6)
}

given intOrd: Ordering[WidgetS] = new Ordering[WidgetS] {
  def compare(a: WidgetS, b: WidgetS): Int = a.size - b.size
}

given String = "anon"

extension (s: String)
  def doubleUp: String = concat(s, s)
  def tripleUp: String = concat(s, concat(s, s))

extension [A](xs: List[A]) def secondOpt: Option[A] = xs.drop(1).headOption

implicit class RichIntS(val i: Int) {
  def twice: Int = i * 2
}

type WidgetList = List[WidgetS]
opaque type Meters = Double

object CallSites {
  def run(w: WidgetS): Unit = {
    helper(1)
    w.render()
    WidgetS(1)
    WidgetS.create().render()
    lowerFactory().chain()
    a.b.method3()
    this.mine()
    super.hashCode()
    "lit".toUpperCase()
    5.toString()
    genericCall[Int](1)
    val n2 = new Ordering[Int] { def compare(p: Int, q: Int): Int = p - q }
    val s1 = s"plain $topVal and ${w.render()} end"
    val e1 = registerCb(handler _)
    val e2 = handler _
    registerCb(handler)
    val r1 = list map transform
    val r2 = 1 :: rest
    w match {
      case ws: WidgetS => use(ws)
      case _ => ()
    }
    for { x <- xs; y <- ys if y > 0 } yield combineXY(x, y)
  }
  def mine(): Int = 1
  def handler(): Unit = ()
  def localHost(): Int = {
    def innerFn(k: Int): Int = k * 2
    val lam = (q: Int) => q + 1
    class LocalClass { def lm(): Int = 3 }
    object LocalObj { def om(): Int = 4 }
    innerFn(lam(1)) + helperCall()
  }
}

object StaticReads {
  def reads(): Unit = {
    val c1 = Registry.count
    val c2 = Http.Ok
    val c3 = com.example.Fq.CONST_READ
    Registry.count = 5
    process(Registry.count)
    Registry.register(null)
  }
}

package object utilpkg {
  def pkgHelper(): Int = 1
  val pkgShared = 2
}
